### DISCUSSION TITLE
Which type to use for an extended Prisma client in generic datastructures like map?

### DISCUSSION THREAD

**Discussion by andreasvh-conceto** [prisma/prisma#18032] [Q&A] [answered]
Hello,

assume i have the following function, which returns an own extended version of the prisma client. In my case appUser is my model, where the return type is extended. Without defining a concrete return type , typing will work out of the box:

```ts
function createExtendedClient(databaseUrl: string) {
    return new PrismaClient({
      datasources: {
        db: {
          url: databaseUrl,
        },
      },
    }).$extends({
      // name: "Testclient",
      result: {
        appUser: {
          fullName: {
            needs: {
              firstname: true,
              lastname: true,
            },
            compute(model) {
              return `${model.firstname} ${model.lastname}`;
            },
          },
        },
      },
    });
  }
}

// this will work directly without defining a  a return type - prisma provides this type out of the box 👍 
createExtendedClient("someUselessUrl").appUser.findFirst().then((u) => u?.fullName)
```
So there seems to be a kind of generic type out of the prisma box, which makes the extended typings available.

I tried now to extend the PrismaClient type like this.

```ts
// this interface is for all the method return types where i would would like to have the the Name property available
export interface IAppUser extends AppUser {
  Name?: string;
}
export type IExtendedPrismaClient = Omit<PrismaClient, "$use">;
function createExtendedClient(databaseUrl: string) : IExtendedPrismaClient 

// here i need to make an ugly cast but even for other method return types, like findAll create 
const user = prismaClient.appUser.findFirst().then((u) => (u as IAppUser).Name);
const userNames = prismaClient.appUser.findMany().then((users) => users.map(u => (u as IAppUser).Name));
```

As you can see i have to perform casts or extend the IExtendedPrismaClient in an ugly way. This issue will be there for every method, which returns the record result like find, findMany, create. The reason why i need such a return type is that i want to store prismaClients in a kind of generic type like a map:

`const customerNameToPrismaClient = new Map<string, IExtendedPrismaClient>`

These kind of datastructures require a specific type and using those i have to perform ugly casts everywhere. I mean - somehow prisma is returning this kind of magic type i need, if i do not specify any return type for the createExtendedClient method. I would like to use the prisma client without performing all these casts like in the first code example.

Any idea, which kind of generic type Prisma is returning here for this extended version of the prismaClient, which can be used in generic datastructures?

Thanks a lot for your help and best regards
Andreas

**Comment by SevInf** [comment, marked as answer]
Hi @andreasvh-conceto.
You can use `ReturnType` helper along with your `createExtendedClient` function to get the derived type for Prisma client with extensions:

```ts
export type ExtendedPrismaClient = ReturnType<typeof createExtendedClient>
```

**Reply by andreasvh-conceto** [reply]
This works perfect. Thanks a lot! :)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
